### PR TITLE
Issue #330 - Projects need controls on cluster-scoped resources

### PR DESCRIPTION
**Pull request by alexmt**
PR introduces ClusterResources field (array of resources represented by group and kind) in app project spec.

If app contains cluster wide resources which is not listed in  ClusterResources field then sync fails and app condition should has error with an appropriate message.

### CODE REVIEW

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -541,6 +546,26 @@ func (spec ApplicationSpec) GetProject() string {
...
 }
 
+func (proj AppProject) IsDefault() bool {
+	return proj.Name == "" || proj.Name == common.DefaultAppProjectName
+}
+
+func (proj AppProject) IsClusterGroupKindPermitted(gk metav1.GroupKind) bool {
+	if proj.IsDefault() {
+		return true
+	}
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types.go]: Actually we no longer treat the default project as anything special, other than it cannot be deleted. So we should remove any special consideration for the default project.

File: pkg/apis/application/v1alpha1/types.go [hunk 2]
```diff
@@ -541,6 +546,26 @@ func (spec ApplicationSpec) GetProject() string {
...
+	for _, item := range proj.Spec.ClusterResources {
+		ok, err := filepath.Match(item.Kind, gk.Kind)
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types.go]: Does this support `*` notation? If not I think we should support this.

File: pkg/apis/application/v1alpha1/types.go [hunk 3]
```diff
@@ -464,6 +466,9 @@ type AppProjectSpec struct {
 	Description string `json:"description,omitempty" protobuf:"bytes,3,opt,name=description"`
 
 	Roles []ProjectRole `json:"roles,omitempty" protobuf:"bytes,4,rep,name=roles"`
+
+	// ClusterResources contains list of cluster level group kinds which are accessible by project apps
+	ClusterResources []metav1.GroupKind `json:"clusterGroupKinds,omitempty" protobuf:"bytes,5,opt,name=clusterGroupKinds"`
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types.go]: Thinking further, how do you think we should handle whitelisting/blacklisting of namespaced resources? For example, especially with the advent of CRDs I believe that eventually, cluster operators will need to restrict what can be permitted even with expensive namespaced resources (e.g. ServiceBroker, DaemonSets, etc...)

File: pkg/apis/application/v1alpha1/types.go (hunk 1 above)
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go]: Got catch! Removed special treatment and added default cluster resource whitelist to the default project.

File: pkg/apis/application/v1alpha1/types.go (hunk 2 above)
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go]: Yes, it is supported by `filepath.Match` function.

File: pkg/apis/application/v1alpha1/types.go (hunk 3 above)
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go]: As we discussed offline renamed ClusterResources to `ClusterResourceWhiteList` and introduced `NamespaceResourceBlackList`.

**Comment by alexmt** [comment]
@jessesuen thank you for review. PTAL

File: server/server.go [hunk 4]
```diff
@@ -125,14 +126,23 @@ func initializeDefaultProject(opts ArgoCDServerOpts) error {
...
 
-	_, err := opts.AppClientset.ArgoprojV1alpha1().AppProjects(opts.Namespace).Create(defaultProj)
-	if apierrors.IsAlreadyExists(err) {
-		return nil
+	existingDefault, err := opts.AppClientset.ArgoprojV1alpha1().AppProjects(opts.Namespace).Get(defaultProj.Name, metav1.GetOptions{})
+	if err == nil {
+		if !reflect.DeepEqual(existingDefault.Spec, defaultProj.Spec) {
+			existingDefault.Spec = defaultProj.Spec
+			_, err = opts.AppClientset.ArgoprojV1alpha1().AppProjects(opts.Namespace).Update(existingDefault)
+		}
```
Comment (jessesuen) [inline server/server.go]: We want to allow users to customize the default project (just like any other project). So we don't want to replace the existing project if it already exists. For users upgrading from v0.8, we'll need need to put in a release note that cluster resources are disabled by default, and let them know the command to allow it.

**Review by jessesuen — CHANGES_REQUESTED** [review]

File: pkg/apis/application/v1alpha1/types.go [hunk 5]
```diff
@@ -541,7 +549,32 @@ func (spec ApplicationSpec) GetProject() string {
 	return spec.Project
 }
 
-// IsSourcePermitted validiates if the provided application's source is a one of the allowed sources for the project.
+func (proj AppProject) IsDefault() bool {
+	return proj.Name == "" || proj.Name == common.DefaultAppProjectName
+}
```
[the remark this answers is not in the document]
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go]: Looks like method is obsoleted. Deleted

File: server/server.go (hunk 4 above)
Comment (alexmt) [reply server/server.go]: Good point. Reverted changes in server.go

File: pkg/apis/application/v1alpha1/types.go [hunk 6]
```diff
@@ -464,6 +466,12 @@ type AppProjectSpec struct {
 	Description string `json:"description,omitempty" protobuf:"bytes,3,opt,name=description"`
 
 	Roles []ProjectRole `json:"roles,omitempty" protobuf:"bytes,4,rep,name=roles"`
+
+	// ClusterResourceWhiteList contains list of white-listed cluster level resources
+	ClusterResourceWhiteList []metav1.GroupKind `json:"clusterResourceWhiteList,omitempty" protobuf:"bytes,5,opt,name=clusterResourceWhiteList"`
+
+	// NamespaceResourceBlackList contains list of black-listed namespace level resources
+	NamespaceResourceBlackList []metav1.GroupKind `json:"namespaceResourceBlackList,omitempty" protobuf:"bytes,6,opt,name=namespaceResourceBlackList"`
```
[the remark this answers is not in the document]
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go]: Fixed.

**Comment by alexmt** [comment]
Thanks for review @jessesuen PTAL

File: cmd/argocd/commands/project.go [hunk 7]
```diff
@@ -677,9 +777,9 @@ func NewProjectListCommand(clientOpts *argocdclient.ClientOptions) *cobra.Comman
 			projects, err := projIf.List(context.Background(), &project.ProjectQuery{})
 			errors.CheckError(err)
 			w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
-			fmt.Fprintf(w, "NAME\tDESCRIPTION\tDESTINATIONS\n")
+			fmt.Fprintf(w, "NAME\tDESCRIPTION\tDESTINATIONS\tSOURCES\tCLUSTER RESOURCE WHITELIST\tNETWORK RESOURCE BLACKLIST\n")
```
Comment (jessesuen) [inline cmd/argocd/commands/project.go]: `NETWORK RESOURCE BLACKLIST` should read `NAMESPACE RESOURCE BLACKLIST`

Also I think these should be hyphenated to indicate it's a single column: `CLUSTER-RESOURCE-WHITELIST` and `NAMESPACE-RESOURCE-BLACKLIST`

File: cmd/argocd/commands/project.go [hunk 8]
```diff
@@ -603,6 +607,102 @@ func NewProjectAddSourceCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+}
+
+// NewProjectAllowClusterResourceCommand returns a new instance of an `argocd proj allow-cluster-resource` command
+func NewProjectAllowClusterResourceCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
+	use := "allow-cluster-resource PROJECT group kind"
+	desc := "Removed cluster wide resource from white list"
+	return modifyProjectResourceCmd(use, desc, clientOpts, func(proj *v1alpha1.AppProject, group string, kind string) bool {
+		for _, item := range proj.Spec.ClusterResourceWhitelist {
+			if item.Group == group && item.Kind == kind {
+				log.Fatalf("Group '%s' and kind '%s' are already whitelisted in project", item.Group, item.Kind)
```
Comment (jessesuen) [inline cmd/argocd/commands/project.go]: Lets make this idempotent.

File: cmd/argocd/commands/project.go [hunk 9]
```diff
@@ -603,6 +607,102 @@ func NewProjectAddSourceCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+}
+
+// NewProjectDenyNamespaceResourceCommand returns a new instance of an `argocd proj deny-namespace-resource` command
+func NewProjectDenyNamespaceResourceCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
+	use := "deny-namespace-resource PROJECT group kind"
+	desc := "Adds namespaced resource to black list"
+	return modifyProjectResourceCmd(use, desc, clientOpts, func(proj *v1alpha1.AppProject, group string, kind string) bool {
+		for _, item := range proj.Spec.NamespaceResourceBlacklist {
+			if item.Group == group && item.Kind == kind {
+				log.Fatalf("Group '%s' and kind '%s' are already blacklisted in project", item.Group, item.Kind)
```
Comment (jessesuen) [inline cmd/argocd/commands/project.go]: Lets make this idempotent.

File: cmd/argocd/commands/project.go [hunk 10]
```diff
@@ -603,6 +607,102 @@ func NewProjectAddSourceCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+	return modifyProjectResourceCmd(use, desc, clientOpts, func(proj *v1alpha1.AppProject, group string, kind string) bool {
+		index := -1
+		for i, item := range proj.Spec.NamespaceResourceBlacklist {
+			if item.Group == group && item.Kind == kind {
+				index = i
+				break
+			}
+		}
+		if index == -1 {
+			log.Info("Specified cluster resource is not black-listed")
```
Comment (jessesuen) [inline cmd/argocd/commands/project.go]: blacklisted

File: pkg/apis/application/v1alpha1/types.go [hunk 11]
```diff
@@ -464,6 +466,12 @@ type AppProjectSpec struct {
 	Description string `json:"description,omitempty" protobuf:"bytes,3,opt,name=description"`
 
 	Roles []ProjectRole `json:"roles,omitempty" protobuf:"bytes,4,rep,name=roles"`
+
+	// ClusterResourceWhitelist contains list of white-listed cluster level resources
+	ClusterResourceWhitelist []metav1.GroupKind `json:"clusterResourceWhitelist,omitempty" protobuf:"bytes,5,opt,name=clusterResourceWhitelist"`
+
+	// NamespaceResourceBlacklist contains list of black-listed namespace level resources
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types.go]: blacklisted

File: pkg/apis/application/v1alpha1/types.go [hunk 12]
```diff
@@ -464,6 +466,12 @@ type AppProjectSpec struct {
 	Description string `json:"description,omitempty" protobuf:"bytes,3,opt,name=description"`
 
 	Roles []ProjectRole `json:"roles,omitempty" protobuf:"bytes,4,rep,name=roles"`
+
+	// ClusterResourceWhitelist contains list of white-listed cluster level resources
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types.go]: whitelisted

**Review by jessesuen — APPROVED** [review]

**Pull request merged**